src/stan_math_backend/Lower_expr.ml

```diff
@@ -483,15 +483,21 @@ and lower_indexed_simple e idcs =
   List.fold idcs ~init:(lower_expr e) ~f:(fun e id ->
       Index (e, idx_minus_one id) )
 
-and lower_expr (Expr.Fixed.{pattern; meta} : Expr.Typed.t) : Cpp.expr =
+and lower_expr ?(promote_reals = false)
+    (Expr.Fixed.{pattern; meta} : Expr.Typed.t) : Cpp.expr =
   let open Exprs in
   match pattern with
   | Var s -> Var s
   | Lit (Str, s) -> literal_string s
   | Lit (Imaginary, s) ->
       fun_call "stan::math::to_complex" [Literal "0"; Literal s]
   | Lit ((Real | Int), s) -> Literal s
-  | Promotion (expr, UReal, _) when is_scalar expr -> lower_expr expr
+  | Promotion (expr, UReal, _) when is_scalar expr ->
+      if promote_reals then
+        (* this can be important for e.g. templated function calls
+           where we might generate an incorrect specification for int *)
+        static_cast Cpp.Double ((lower_expr ~promote_reals:true) expr)
```

Comment:
The inner expression does not need `~promote_reals:true` unless you want to promote the value twice (normal code should not have nested promotions but it can happen after optimization, e.g. if you add `~promote_reals:true` to the `FnMakeArray` code on line 401 some tests in `test/integration/good/compiler-optimizations` produce
`static_cast<double>(static_cast<double>(1000000))`

Reply:
Fair enough (though I assume that the C++ compiler is smart enough to delete the outer one)

Reply:
If we were to rely on the C++ compiler to ignore superfluous casts (which we certainly can do) then it would be simpler to ignore `promote_reals` completely and generate a `static_cast` for every `Promotion`. I commented because this `~promote_reals:true` looks very much intentional but cannot really accomplish anything.

Reply:
Yeah, I went for the change which would fix the bug and result in minimal changes in the generated source. `static_cast`ing everything is certainly an option, I confirmed looking at godbolt that it generates identical assembly as the "automatic" upcasting C++ normally does.